### ISSUE TITLE
Many issues on zeppelin

### ISSUE THREAD

**Issue by quancore** [dask/dask-yarn#64]
I am using aws emr with 1 master and 2 worker node (r5.xlarge). Hadoop, spark, and Zeppelin is all set up. I am using zeppelin for notebook processing.  My bootstrap script for emr:
```
#!/usr/bin/env bash
set -x -e

#  move /usr/local to /mnt/usr-moved/local; else run out of space on /
sudo mkdir /mnt/usr-moved
sudo mv /usr/local /mnt/usr-moved/
sudo ln -s /mnt/usr-moved/local /usr/
sudo mv /usr/share /mnt/usr-moved/
sudo ln -s /mnt/usr-moved/share /usr/


#sudo bash -c 'yum update -y && yum upgrade && sudo yum groupinstall -y 'Development Tools''
sudo bash -c 'yum group install -y 'Development Tools''
sudo yum install -y htop

# Install anaconda
wget https://repo.continuum.io/archive/Anaconda3-5.3.0-Linux-x86_64.sh -O $HOME/anaconda.sh && /bin/bash ~/anaconda.sh -b -p $HOME/anaconda
echo -e '\nexport PATH=$HOME/anaconda/bin:$PATH' >> $HOME/.bashrc && source $HOME/.bashrc


conda config --set always_yes yes --set changeps1 no
conda config -f --add channels conda-forge
conda config -f --add channels defaults
conda update conda
conda update --all


# cleanup
rm ~/anaconda.sh

echo bootstrap_conda.sh completed. PATH now: $PATH
export PYSPARK_PYTHON="$HOME/anaconda/bin/python3.7"

# setup git
sudo yum install -y git

# install additional third party libraries
echo "Installing base packages"
conda install \
-c defaults \
-c conda-forge \
-c anaconda \
-y \
-q \
dask-yarn>=0.4.1 \
pyarrow \
s3fs \
nomkl \
conda-pack \
tensorflow \
grpcio \
protobuf \

yes | pip install fancyimpute
yes | pip install -U pandasql

if grep isMaster /mnt/var/lib/info/instance.json | grep true;
then
echo "Packaging environment"
conda pack -q -o $HOME/environment.tar.gz

echo "Configuring Dask"
mkdir -p $HOME/.config/dask
cat <<EOT >> $HOME/.config/dask/config.yaml
distributed:
  dashboard:
    link: "/proxy/{port}/status"
yarn:
  environment: /home/hadoop/environment.tar.gz
  worker:
    env:
      ARROW_LIBHDFS_DIR: /usr/lib/hadoop/lib/native/
  client:
    env:
      ARROW_LIBHDFS_DIR: /usr/lib/hadoop/lib/native/
EOT
# Also set ARROW_LIBHDFS_DIR in ~/.bashrc so it's set for the local user
echo -e '\nexport ARROW_LIBHDFS_DIR=/usr/lib/hadoop/lib/native' >> $HOME/.bashrc

fi
```

I have followed your aws emr guide to set up my environment however there are many problems:

> First start with a question, I am using anaconda instead of miniconda, and it gives 2GiB conda env file. Is It okay?

1) When I try to set up a cluster using local mode like:
```
import dask
import dask.distributed
from dask_yarn import YarnCluster
from dask.distributed import Client
from sklearn.externals import joblib

cluster = YarnCluster(deploy_mode='local')
```
it gives me:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-8bae7fd69e7b> in <module>
      1 # Create a cluster in local deploy mode, to have access to the dashboard
----> 2 cluster = YarnCluster(deploy_mode='local')
      3 # cluster = YarnCluster.from_specification('/home/hadoop/.config/dask/config.yaml')
      4 # cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz', worker_vcores=4, worker_memory='4 GiB')
      5 

/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py in __init__(self, environment, n_workers, worker_vcores, worker_memory, worker_restarts, worker_env, scheduler_vcores, scheduler_memory, deploy_mode, name, queue, tags, user, skein_client)
    291                                    queue=queue,
    292                                    tags=tags,
--> 293                                    user=user)
    294 
    295         self._start_cluster(spec, skein_client)

/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py in _make_specification(**kwargs)
    116                "\n"
    117                "See http://yarn.dask.org/environments.html for more information.")
--> 118         raise ValueError(msg)
    119 
    120     n_workers = lookup(kwargs, 'n_workers', 'yarn.worker.count')

ValueError: You must provide a path to a Python environment for the workers.
This may be one of the following:
- A conda environment archived with conda-pack
- A virtual environment archived with venv-pack
- A path to a conda environment, specified as conda://...
- A path to a virtual environment, specified as venv://...
- A path to a python binary to use, specified as python://...

See http://yarn.dask.org/environments.html for more information.
```
After I tried to go with:
```
cluster = YarnCluster.from_specification('/home/hadoop/.config/dask/config.yaml')
```
gives me:
```
ValueError: Unknown extra keys for ApplicationSpec:
- distributed
- yarn
```
my config.yaml:
```
distributed:
  dashboard:
    link: "/proxy/{port}/status"
yarn:
  environment: /home/hadoop/environment.tar.gz
  worker:
    env:
      ARROW_LIBHDFS_DIR: /usr/lib/hadoop/lib/native/
  client:
    env:
      ARROW_LIBHDFS_DIR: /usr/lib/hadoop/lib/native/
```
when I do :
```
cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz', worker_vcores=4, worker_memory='4 GiB')
```
It just waiting, and running but no ends.

Eventually, I can do:
```
cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz')
cluster.scale(10)
```
It works okay. But I cannot configure worker memory, vcore etc.

2) When I try to reach dask dashboard by printing:
```
print(cluster.dashboard_link)
```
It gives me `None`
3) Even if I continue with these problems, when I run grid search like:
```
with joblib.parallel_backend('dask'):
   # do grid search
```
When I connect worker nodes with ssh and check resource utilization, they are idle. Only the master node with 2 cores is executing.

**Comment by quancore** [comment]
Update for the issue:
I managed to make a cluster by the code:
```
cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz', n_workers=4, worker_vcores=4, worker_memory='4 GiB')
client = Client(cluster)
print(client)
print(cluster.dashboard_link)
print(client.ncores())
```

Output:

> <Client: scheduler='tcp://172.31.28.103:39217' processes=4 cores=16>
None
{'tcp://172.31.28.103:34907': 4,
 'tcp://172.31.28.103:37501': 4,
 'tcp://172.31.28.103:46285': 4,
 'tcp://172.31.28.103:46391': 4}

After creation, I am trying to use backend by running: 
```
# non-linear knn
model = 'KNeighbors'

param_grid = {'n_neighbors':np.arange(3,15,4)}
# param_grid = {}
model_dict[model] = KNeighborsRegressor()
with joblib.parallel_backend('dask'):
    model_dict[model], cv_score, grid_results = data_util.train_model(model_dict[model], X_train, y_train, X_test, y_test, regressor_metric,
                                                                  splits=cv, repeats=repeats,
                                                                  n_jobs=n_jobs,
                                                                  verbosity=verbosity,
                                                                  ignore_war=True)
cv_score['name'] = model
if model not in score_models.index:
    score_models = score_models.append(cv_score.set_index('name', append=True).swaplevel(0, 1), sort=False)
```
train_model method:
```
def train_model(model, X_train, y_train, X_test, y_test, scorers,
                param_grid=[],
                refit='neg_mean_squared_error',
                splits=5, repeats=5,
                n_jobs=4,
                verbosity=1,
                ignore_war=True):
    """Train a model with given parameters.

    Parameters
    ----------
    model : `class`
        Model to be fitted.

    train_data : `dataframe`
        Dict of X_train, y_train dataframes of observed and target train variables.

    test_data: `dataframe`
        Dict of X_test, y_test dataframes of observed and target test variables.

    scorers : `list`
        List of scorer metrics.

    refit : `string`
        Which metrics will be used during refit of cv search.

    param_grid : `list`
        List of parameters will be used to train the model.

    splits : `int`
        Cross-validation split count.

    repeats : `int`
        Number of repeats for cv.

    n_jobs : `int`
        Number of paralell jobs.

    verbosity : `int`
        Log verbosity level of the model.

    ignore_war : `bool`
        Flag indicating whether we ignore warnings.

    Returns
    -------
    nonlabel_cols : `list`
        List of columns filtered.
    """
    # define vars
    cv_mean, cv_std, test_scores = {}, {}, {}
    # get train and test data
    # X_train, y_train = train_data[0], train_data[1]
    # X_test, y_test = test_data[0], test_data[1]
    flat_y_train = y_train.values.ravel()
    flat_y_test = y_test.values.ravel()

    # create cross-validation method
    rkfold = RepeatedKFold(n_splits=splits, n_repeats=repeats)

    # perform a grid / random search if param_grid given
    if len(param_grid) > 0:
        with warnings.catch_warnings():
            if ignore_war:
                # ignore all caught warnings
                warnings.filterwarnings("ignore")
            else:
                warnings.filterwarnings("default")

            print('Hyperparameter tuning is beginning...')
            # setup grid search parameters
            # gsearch = GridSearchCV(model, param_grid, cv=rkfold,
            #                       scoring=regressor_metric,
            #                       verbose=verbosity, return_train_score=True,
            #                       n_jobs=n_jobs)
            gsearch = RandomizedSearchCV(model, param_grid, cv=rkfold,
                                         scoring=scorers,
                                         refit=refit,
                                         verbose=verbosity,
                                         return_train_score=True,
                                         n_jobs=n_jobs)

            # search the grid
            gsearch.fit(X_train, flat_y_train)

        # extract best model from the grid
        model = gsearch.best_estimator_
        best_idx = gsearch.best_index_

        # get cv-scores for best model
        grid_results = pd.DataFrame(gsearch.cv_results_)

        for scorer in scorers:
            cv_mean[scorer] = abs(grid_results.loc[best_idx, 'mean_test_%s' % (scorer)])
            cv_std[scorer] = grid_results.loc[best_idx, 'std_test_%s' % (scorer)]

    # no grid / random search, just cross-val score for given model
    else:
        grid_results = []
        print('Fitting model...')
        model.fit(X_train, flat_y_train)
        print('Getting cros. val score...')
        cv_results = cross_validate(model, X_train, flat_y_train,
                                    scoring=scorers,
                                    n_jobs=n_jobs,
                                    cv=rkfold,
                                    verbose=verbosity)
        for scorer in scorers:
            res_scores = cv_results['test_%s' % (scorer)]
            cv_mean[scorer] = abs(np.mean(res_scores))
            cv_std[scorer] = np.std(res_scores)

    # predict y using the fitted model
    y_test_pred = model.predict(X_test)
    for scorer in scorers:
        scorer_method = get_scorer(scorer)
        # get test score
        test_scores[scorer] = abs(scorer_method(model, X_test, flat_y_test))

    # combine mean and std cv-score in to a pandas dataframe
    cv_score = pd.DataFrame.from_records([cv_mean, cv_std, test_scores])
    cv_score.index = ['cv_mean_score', 'cv_std', 'test_score']

    # print stats on model performance
    print('----------------------')
    print(model)
    for scorer in scorers:
        print('------ {} ----------'.format(scorer))
        print('test_score=', test_scores[scorer])
        print('cross_val: mean=', cv_mean[scorer], ', std=', cv_std[scorer])

    # residual plots
    y_test_pred = pd.Series(y_test_pred.ravel(), index=y_test.index)
    resid = y_test['mu'] - y_test_pred
    mean_resid = resid.mean()
    std_resid = resid.std()
    z = (resid - mean_resid)/std_resid
    n_outliers = sum(abs(z) > 3)

    plt.figure(figsize=(15, 5))
    ax_131 = plt.subplot(1, 3, 1)
    plt.plot(y_test, y_test_pred, '.')
    plt.xlabel('y_test')
    plt.ylabel('y_pred');
    plt.title('corr = {:.3f}'.format(np.corrcoef(y_test['mu'], y_test_pred)[0][1]))

    ax_132 = plt.subplot(1, 3, 2)
    plt.plot(y_test_pred, resid, '.')
    plt.xlabel('y_pred')
    plt.ylabel('y_test - y_pred');
    plt.title('std resid = {:.3f}'.format(std_resid))

    ax_133 = plt.subplot(1, 3, 3)
    z.plot.hist(bins=50, ax=ax_133)
    plt.xlabel('z')
    plt.title('{:.0f} samples with z>3'.format(n_outliers))

    return model, cv_score, grid_results
```
Which is basically running a sklearn grid search.
The output of execution:

```
CancelledError                            Traceback (most recent call last)
/home/hadoop/anaconda/lib/python3.7/site-packages/sklearn/externals/joblib/parallel.py in retrieve(self)
    834                 else:
--> 835                     self._output.extend(job.get())
    836 

/home/hadoop/anaconda/lib/python3.7/site-packages/sklearn/externals/joblib/_dask.py in get()
    229         def get():
--> 230             return ref().result()
    231 

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/client.py in result(self, timeout)
    228         elif self.status == "cancelled":
--> 229             raise result
    230         else:

CancelledError: _fit_and_score-batch-7858d9c2cd274fbaadb7de50e34af32a

During handling of the above exception, another exception occurred:

CommClosedError                           Traceback (most recent call last)
<ipython-input-16-3bae4f545c43> in <module>
     10                                                                   n_jobs=n_jobs,
     11                                                                   verbosity=verbosity,
---> 12                                                                   ignore_war=True)
     13 cv_score['name'] = model
     14 if model not in score_models.index:

/home/hadoop/cto-esports-ranking-prediction/data_analysis/utils.py in train_model(model, X_train, y_train, X_test, y_test, scorers, param_grid, refit, splits, repeats, n_jobs, verbosity, ignore_war)
    240                                     n_jobs=n_jobs,
    241                                     cv=rkfold,
--> 242                                     verbose=verbosity)
    243         for scorer in scorers:
    244             res_scores = cv_results['test_%s' % (scorer)]

/home/hadoop/anaconda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py in cross_validate(estimator, X, y, groups, scoring, cv, n_jobs, verbose, fit_params, pre_dispatch, return_train_score, return_estimator, error_score)
    238             return_times=True, return_estimator=return_estimator,
    239             error_score=error_score)
--> 240         for train, test in cv.split(X, y, groups))
    241 
    242     zipped_scores = list(zip(*scores))

/home/hadoop/anaconda/lib/python3.7/site-packages/sklearn/externals/joblib/parallel.py in __call__(self, iterable)
    928 
    929             with self._backend.retrieval_context():
--> 930                 self.retrieve()
    931             # Make sure that we get a last message telling us we are done
    932             elapsed_time = time.time() - self._start_time

/home/hadoop/anaconda/lib/python3.7/site-packages/sklearn/externals/joblib/parallel.py in retrieve(self)
    853                     # scheduling.
    854                     ensure_ready = self._managed_backend
--> 855                     backend.abort_everything(ensure_ready=ensure_ready)
    856 
    857                 if isinstance(exception, TransportableException):

/home/hadoop/anaconda/lib/python3.7/site-packages/sklearn/externals/joblib/_dask.py in abort_everything(self, ensure_ready)
    238         joblib.Parallel will never access those results
    239         """
--> 240         self.client.cancel(self.task_futures)
    241         self.task_futures.clear()
    242 

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/client.py in cancel(self, futures, asynchronous, force)
   2095             Cancel this future even if other clients desire it
   2096         """
-> 2097         return self.sync(self._cancel, futures, asynchronous=asynchronous, force=force)
   2098 
   2099     @gen.coroutine

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/client.py in sync(self, func, *args, **kwargs)
    751             return future
    752         else:
--> 753             return sync(self.loop, func, *args, **kwargs)
    754 
    755     def __repr__(self):

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, *args, **kwargs)
    329             e.wait(10)
    330     if error[0]:
--> 331         six.reraise(*error[0])
    332     else:
    333         return result[0]

/home/hadoop/anaconda/lib/python3.7/site-packages/six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/utils.py in f()
    314             if timeout is not None:
    315                 future = gen.with_timeout(timedelta(seconds=timeout), future)
--> 316             result[0] = yield future
    317         except Exception as exc:
    318             error[0] = sys.exc_info()

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    727 
    728                     try:
--> 729                         value = future.result()
    730                     except Exception:
    731                         exc_info = sys.exc_info()

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    734                     if exc_info is not None:
    735                         try:
--> 736                             yielded = self.gen.throw(*exc_info)  # type: ignore
    737                         finally:
    738                             # Break up a reference to itself

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/client.py in _cancel(self, futures, force)
   2075     def _cancel(self, futures, force=False):
   2076         keys = list({tokey(f.key) for f in futures_of(futures)})
-> 2077         yield self.scheduler.cancel(keys=keys, client=self.id, force=force)
   2078         for k in keys:
   2079             st = self.futures.pop(k, None)

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    727 
    728                     try:
--> 729                         value = future.result()
    730                     except Exception:
    731                         exc_info = sys.exc_info()

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    734                     if exc_info is not None:
    735                         try:
--> 736                             yielded = self.gen.throw(*exc_info)  # type: ignore
    737                         finally:
    738                             # Break up a reference to itself

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/core.py in send_recv_from_rpc(**kwargs)
    737             name, comm.name = comm.name, "ConnectionPool." + key
    738             try:
--> 739                 result = yield send_recv(comm=comm, op=key, **kwargs)
    740             finally:
    741                 self.pool.reuse(self.addr, comm)

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    727 
    728                     try:
--> 729                         value = future.result()
    730                     except Exception:
    731                         exc_info = sys.exc_info()

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    734                     if exc_info is not None:
    735                         try:
--> 736                             yielded = self.gen.throw(*exc_info)  # type: ignore
    737                         finally:
    738                             # Break up a reference to itself

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/core.py in send_recv(comm, reply, serializers, deserializers, **kwargs)
    531         yield comm.write(msg, serializers=serializers, on_error="raise")
    532         if reply:
--> 533             response = yield comm.read(deserializers=deserializers)
    534         else:
    535             response = None

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    727 
    728                     try:
--> 729                         value = future.result()
    730                     except Exception:
    731                         exc_info = sys.exc_info()

/home/hadoop/anaconda/lib/python3.7/site-packages/tornado/gen.py in run(self)
    734                     if exc_info is not None:
    735                         try:
--> 736                             yielded = self.gen.throw(*exc_info)  # type: ignore
    737                         finally:
    738                             # Break up a reference to itself

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/comm/tcp.py in read(self, deserializers)
    212             self.stream = None
    213             if not shutting_down():
--> 214                 convert_stream_closed_error(self, e)
    215         else:
    216             try:

/home/hadoop/anaconda/lib/python3.7/site-packages/distributed/comm/tcp.py in convert_stream_closed_error(obj, exc)
    139         raise CommClosedError("in %s: %s: %s" % (obj, exc.__class__.__name__, exc))
    140     else:
--> 141         raise CommClosedError("in %s: %s" % (obj, exc))
    142 
    143 

CommClosedError: in <closed TCP>: Stream is closed  
```
And after this point cluster become `not connected` state.
I suspect that the issue can be related the [issue](https://github.com/dask/dask/issues/4511)

**Comment by quancore** [comment]
Bump it.

**Comment by TomAugspurger** [comment]
This issue is hard for me to follow. Could you summarize what the problem is, perhaps with a minimal example? http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

Is all the code in https://github.com/dask/dask-yarn/issues/64#issuecomment-491524586 necessary (plotting, etc.)?

**Comment by jcrist** [comment]
I agree with Tom above, this is a lot of different issues here. Good issues are concise and reproducible, please refer to https://blog.dask.org/2018/02/28/minimal-bug-reports for more information. I'll attempt to respond to a few of the issues inline below.

---

> First start with a question, I am using anaconda instead of miniconda, and it gives 2GiB conda env file. Is It okay?

Yes, this is fine, but not ideal, 2 GiB is a lot to move around. Alternatively you could have a bootstrap script install anaconda on every node (workers included), and specify `environment=python:///path/to/bin/python` to the constructor instead.

> When I try to set up a cluster using local mode like...

This looks like your dask configuration isn't being picked up. This may be due to installation issues, or zeppelin changing your environment (likely running as a different user) so the files aren't found. I recommend checking the following:

1. Look in the loaded dask configuration to see if any yarn configuration values are loaded.

```python
import dask
# Should contain some yarn configuration values
print(dask.config.config)
```

2. Look at the configuration file to see if you have access, and that the file is the same (zeppelin may run your kernel as a different user, or on a different machine, not sure, I have no experience with zeppelin):

```python
# Should be readable, and have the info from the bootstrap script
with open('~/.config/dask/config.yaml') as f:
    print(f.read())
```

> After I tried to go with:
> `cluster = YarnCluster.from_specification('/home/hadoop/.config/dask/config.yaml')`

`YarnCluster.from_specification` is for loading a [skein specification](https://jcrist.github.io/skein/specification.html), and has nothing to do with the dask config file here. Please see the documentation.

> when I do :
> `cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz', worker_vcores=4, worker_memory='4 GiB')`
> It just waiting, and running but no ends.

this may be due to limits on your yarn cluster. YARN may not allow workers of that size (this is `yarn.scheduler.maximum-allocation-vcores` and `yarn.scheduler.maximum-allocation-mb` in `yarn-site.xml`), or your cluster nodes may not support workers of that size. Or it could be something else entirely. From your second comment it looks like you got things working with similar configuration at one point, so this also may be an EMR issue. If you continue to have problems, seeing the application logs would be helpful here:

```bash
$ yarn logs -applicationId <your-application-id>
```

> Eventually, I can do:
> 
> ```python
> cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz')
> cluster.scale(10)
> ```
>
> When I try to reach dask dashboard by printing:
> 
> `print(cluster.dashboard_link)`
>
> It gives me None

The dashboard is only available if bokeh is installed, which may not be true given your setup above. Further, the dashboard runs on the scheduler and may not be accessible through the firewall unless you specify `deploy_mode='local'`.

---

The grid search issues are unrelated to dask-yarn, please make a new issue in dask-ml, or continue discussion in https://github.com/dask/dask/issues/4511.

**Comment by quancore** [comment]
> python:

First of all, thank you for your support. I started with your first suggestion about env. I have run the command:
```
cluster = YarnCluster(environment='python:/home/hadoop/anaconda/bin/python', n_workers=4, worker_vcores=2, worker_memory='2 GiB')

```
Output:
```
19/05/15 09:35:32 INFO client.RMProxy: Connecting to ResourceManager at ip-172-31-11-209.eu-central-1.compute.internal/172.31.11.209:8032
19/05/15 09:35:42 INFO skein.Driver: Driver started, listening on 40621
19/05/15 09:35:46 INFO skein.Driver: Uploading application resources to hdfs://ip-172-31-11-209.eu-central-1.compute.internal:8020/user/zeppelin/.skein/application_1557838259721_0006
19/05/15 09:35:49 INFO skein.Driver: Submitting application...
19/05/15 09:35:50 INFO impl.YarnClientImpl: Submitted application application_1557838259721_0006
19/05/15 09:36:55 INFO impl.YarnClientImpl: Killed application application_1557838259721_0006

Traceback (most recent call last):
  File "/tmp/zeppelin_python-4478919184514491789.py", line 312, in <module>
    exec(code, _zcUserQueryNameSpace)
  File "<stdin>", line 1, in <module>
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py", line 295, in __init__
    self._start_cluster(spec, skein_client)
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py", line 375, in _start_cluster
    scheduler_address = app.kv.wait('dask.scheduler').decode()
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/skein/kv.py", line 655, in wait
    event = event_queue.get()
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/skein/kv.py", line 281, in get
    raise out
skein.exceptions.ConnectionError: Unable to connect to application

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/zeppelin_python-4478919184514491789.py", line 319, in <module>
    raise Exception(traceback.format_exc())
Exception: Traceback (most recent call last):
  File "/tmp/zeppelin_python-4478919184514491789.py", line 312, in <module>
    exec(code, _zcUserQueryNameSpace)
  File "<stdin>", line 1, in <module>
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py", line 295, in __init__
    self._start_cluster(spec, skein_client)
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py", line 375, in _start_cluster
    scheduler_address = app.kv.wait('dask.scheduler').decode()
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/skein/kv.py", line 655, in wait
    event = event_queue.get()
  File "/home/hadoop/anaconda/lib/python3.7/site-packages/skein/kv.py", line 281, in get
    raise out
skein.exceptions.ConnectionError: Unable to connect to application
```
How can I handle this step to proceed next steps?

**Comment by quancore** [comment]
When I rerun the cell below:
```
cluster = YarnCluster(environment='python:/home/hadoop/anaconda/bin/python', n_workers=4, worker_vcores=2, worker_memory='2 GiB')
```
I do not even get the error above:
```
---------------------------------------------------------------------------
ConnectionError                           Traceback (most recent call last)
<ipython-input-4-430481e6fe19> in <module>
      2 # cluster = YarnCluster(deploy_mode='local')
      3 # cluster = YarnCluster.from_specification('/home/hadoop/.config/dask/config.yaml')
----> 4 cluster = YarnCluster(environment='python:/home/hadoop/anaconda/bin/python', n_workers=4, worker_vcores=2, worker_memory='2 GiB')
      5 # cluster = YarnCluster(environment='/home/hadoop/environment.tar.gz')

/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py in __init__(self, environment, n_workers, worker_vcores, worker_memory, worker_restarts, worker_env, scheduler_vcores, scheduler_memory, deploy_mode, name, queue, tags, user, skein_client)
    293                                    user=user)
    294 
--> 295         self._start_cluster(spec, skein_client)
    296 
    297     @cached_property

/home/hadoop/anaconda/lib/python3.7/site-packages/dask_yarn/core.py in _start_cluster(self, spec, skein_client)
    373             app = skein_client.submit_and_connect(spec)
    374             try:
--> 375                 scheduler_address = app.kv.wait('dask.scheduler').decode()
    376                 dashboard_address = app.kv.get('dask.dashboard')
    377                 if dashboard_address is not None:

/home/hadoop/anaconda/lib/python3.7/site-packages/skein/kv.py in wait(self, key, return_owner)
    653                 return res
    654 
--> 655             event = event_queue.get()
    656 
    657         return event.result if return_owner else event.result.value

/home/hadoop/anaconda/lib/python3.7/site-packages/skein/kv.py in get(self, block, timeout)
    279         if isinstance(out, Exception):
    280             self._exception = out
--> 281             raise out
    282         return out
    283 

ConnectionError: Unable to connect to application
```

**Comment by jcrist** [comment]
Apologies for the delayed response here. When a cluster starts it's usually necessary to get the application logs to see what happened. Can you reply with them here:

```
$ yarn logs -applicationId application application_1557838259721_0006
```

**Comment by jcrist** [comment]
Closing as stale. Feel free to reopen if the issue persists.
